include/exadg/fluid_structure_interaction/precice/precice_adapter.h

```diff
@@ -0,0 +1,380 @@
+
+  AssertThrow(dim == precice->getDimensions(), ExcInternalError());
+  AssertThrow(dim > 1, ExcNotImplemented());
+}
+
+
+
+template<int dim, int data_dim, typename VectorType, typename VectorizedArrayType>
+void
+Adapter<dim, data_dim, VectorType, VectorizedArrayType>::add_write_surface(
+  const types::boundary_id                                            dealii_boundary_surface_id,
+  const std::string &                                                 mesh_name,
+  const std::vector<std::string> &                                    write_data_names,
+  const WriteDataType                                                 write_data_type,
+  std::shared_ptr<const MatrixFree<dim, double, VectorizedArrayType>> data,
+  const unsigned int                                                  dof_index,
+  const unsigned int                                                  quad_index)
+{
+  // Check, if we already have such an interface
+  const auto found_reader = reader.find(mesh_name);
+
+  if(found_reader != reader.end())
+  {
+    // Duplicate the shared pointer
+    writer.insert({mesh_name, found_reader->second});
+  }
+  else if(write_data_type == WriteDataType::values_on_dofs)
+  {
+    writer.insert({mesh_name,
+                   std::make_shared<DoFSurface<dim, data_dim, VectorizedArrayType>>(
+                     data, precice, mesh_name, dealii_boundary_surface_id, dof_index)});
+  }
+  else
+  {
+    Assert(write_data_type == WriteDataType::values_on_q_points ||
+             write_data_type == WriteDataType::normal_gradients_on_q_points,
+           ExcNotImplemented());
+    writer.insert({mesh_name,
+                   std::make_shared<QuadSurface<dim, data_dim, VectorizedArrayType>>(
+                     data, precice, mesh_name, dealii_boundary_surface_id, dof_index, quad_index)});
+  }
+
+  // Register the write data
+  const std::vector<Point<dim>> points;
+  for(const auto & data_name : write_data_names)
+    writer.at(mesh_name)->add_write_data(data_name);
+
+  writer.at(mesh_name)->set_write_data_type(write_data_type);
+  // Finally initialize the surface
+  writer.at(mesh_name)->define_coupling_mesh(points);
+}
+
+
+
+template<int dim, int data_dim, typename VectorType, typename VectorizedArrayType>
+void
+Adapter<dim, data_dim, VectorType, VectorizedArrayType>::add_read_surface(
+  const std::vector<Point<dim>> &                                      points,
+  std::shared_ptr<const MatrixFree<dim, double, VectorizedArrayType>>  data,
+  std::shared_ptr<ExaDG::preCICE::InterfaceCoupling<dim, dim, double>> exadg_terminal,
+  const std::string &                                                  mesh_name,
+  const std::vector<std::string> &                                     read_data_names)
+{
+  reader.insert(
+    {mesh_name,
+     std::make_shared<ExaDGSurface<dim, data_dim, VectorizedArrayType>>(data, precice, mesh_name)});
```

Comment:
I would create the wrapper for ExaDG outside of this function to not introduce here a dependency to ExaDG.